### PR TITLE
feat: Add support for svg sprite

### PR DESCRIPTION
**Pull request by paulomfr**
## Discussion
Issue: #11 

## What this PR does
Add support for sprites svg, when add a svg file in `/sprite-svg` folder, webpack will bundle in one file `sprite-svg.svg` and the PHP will include the svg content after body tag opened.

### CODE REVIEW

File: app/helpers/assets.php [hunk 1]
```diff
@@ -74,6 +74,21 @@ function app_action_theme_enqueue_assets() {
 	App::theme()->assets()->enqueueStyle( 'theme-styles', get_template_directory_uri() . '/style.css' );
 }
 
+/**
+ * Enqueue sprite svg.
+ *
+ * @return void
+ */
+function app_enqueue_svg_sprite() {
```
Comment (atanas-dev) [inline app/helpers/assets.php]:
```suggestion
function app_load_svg_sprite() {
```

File: app/helpers/assets.php [hunk 2]
```diff
@@ -74,6 +74,21 @@ function app_action_theme_enqueue_assets() {
...
+	$file = APP_DIST_DIR . 'images/sprite-svg.svg';
```
Comment (atanas-dev) [inline app/helpers/assets.php]:
```suggestion
	$file = APP_DIST_DIR . 'images' . DIRECTORY_SEPARATOR . 'sprite.svg';
```

File: app/helpers/assets.php [hunk 3]
```diff
@@ -74,6 +74,21 @@ function app_action_theme_enqueue_assets() {
...
+
+	if ( ! file_exists( $file ) ) {
+		return;
+	}
+
+	include $file;
```
Comment (atanas-dev) [inline app/helpers/assets.php]:
```suggestion
	readfile( $file );
```

File: app/hooks.php [hunk 4]
```diff
@@ -18,6 +18,7 @@
 /**
  * Assets
  */
+add_action( 'wp_body_open', 'app_enqueue_svg_sprite' );
```
Comment (atanas-dev) [inline app/hooks.php]: This means that the actual site content will be delayed until we serve the SVG sprite - this should be moved to wp_footer instead as glyphs are complimentary to your site, not critical above-the-fold content.

File: app/helpers/assets.php [hunk 5]
```diff
@@ -74,6 +74,21 @@ function app_action_theme_enqueue_assets() {
 	App::theme()->assets()->enqueueStyle( 'theme-styles', get_template_directory_uri() . '/style.css' );
 }
 
+/**
+ * Enqueue sprite svg.
```
Comment (atanas-dev) [inline app/helpers/assets.php]:
```suggestion
 * Load SVG sprite.
```

File: composer.json [hunk 6]
```diff
@@ -20,16 +20,30 @@
       "role": "Developer"
     }
   ],
+  "repositories": [
```
Comment (atanas-dev) [inline composer.json]: All changes to this file are not relevant to the PR and should be removed from it.

File: composer.lock [hunk 7]
```diff
@@ -0,0 +1,1155 @@
+{
```
Comment (atanas-dev) [inline composer.lock]: This file is intentionally not included with the theme and should not be included in this PR.

File: package.json [hunk 8]
```diff
@@ -50,6 +50,8 @@
     "postcss-loader": "^2.1.4",
     "sass-loader": "^7.0.1",
     "style-loader": "^0.21.0",
+    "svg-sprite-loader": "^4.2.5",
+    "terser-webpack-plugin": "^2.3.4",
```
Comment (atanas-dev) [inline package.json]: Please merge latest master in your branch - the terser plugin was removed from master but your PR is bringing it back.

File: package.json [hunk 9]
```diff
@@ -70,6 +72,7 @@
     "stylelint-config-recommended-scss": "^3.2.0",
     "stylelint-order": "^0.8.1",
     "stylelint-scss": "^3.1.0",
+    "svgo-loader": "^2.2.1"
```
Comment (atanas-dev) [inline package.json]: If this dependency is needed for the assets to be built it should be included as a dependency rather than a devDependency.

File: resources/build/lib/utils.js [hunk 10]
```diff
@@ -52,7 +52,8 @@ module.exports.distFontsPath = destPath =>
 module.exports.tests = {
   scripts: /\.(js|jsx)$/,
   styles: /\.(css|scss|sass)$/,
-  images: /(resources|dist|node_modules)[\\/].*\.(ico|jpg|jpeg|png|svg|gif)$/,
+  svgs: /(resources|dist|node_modules)[\\/]sprite-svg.*\.(ico|jpg|jpeg|png|svg|gif)$/,
```
Comment (atanas-dev) [inline resources/build/lib/utils.js]: The allowed extensions should only be limited to `.svg`

File: resources/build/webpack.development.js [hunk 11]
```diff
@@ -55,6 +56,8 @@ const plugins = [
     jQuery: 'jquery',
   }),
   spriteSmith,
+  spriteSvg,
+  browsersync,
```
Comment (atanas-dev) [inline resources/build/webpack.development.js]: Similar as above - this was removed in master but is included as a change in your PR - please merge latest master into your branch.

File: resources/build/webpack.development.js [hunk 12]
```diff
@@ -144,6 +147,34 @@ module.exports = {
         ],
       },
 
+      /**
+       * Handle svg sprites.
```
Comment (atanas-dev) [inline resources/build/webpack.development.js]:
```suggestion
       * Handle SVG sprites.
```

File: package.json (hunk 9 above)
Comment (atanas-dev) [inline package.json]: The svgo package docs state that `svgo` should be installed explicitly as well.

File: resources/build/webpack.development.js [hunk 13]
```diff
@@ -144,6 +147,34 @@ module.exports = {
...
+              spriteFilename: 'images/sprite-svg.svg',
+            },
+          },
+          {
+            loader: 'svgo-loader',
+            options: {
+              plugins: [
+                {
+                  removeAttrs: {
+                    attrs: '*:(stroke|fill)*',
```
Comment (atanas-dev) [inline resources/build/webpack.development.js]: Could you please explain the need for this attribute filter?

File: resources/images/sprite-svg/angle-right.svg [hunk 14]
```diff
@@ -0,0 +1,3 @@
+<svg viewBox="0 0 11 32">
```
Comment (atanas-dev) [inline resources/images/sprite-svg/angle-right.svg]: This test file should not be included in the PR.

File: resources/scripts/frontend/index.js [hunk 15]
```diff
@@ -6,5 +6,7 @@ import '@images/favicon.ico';
 import 'airbnb-browser-shims';
 // Uncomment the following line to enable lazy-loading of media images:
 // import './lazyload';
+// Uncomment the following line to enable sprite svg:
```
Comment (atanas-dev) [inline resources/scripts/frontend/index.js]: If there are no svgs used and a sprite is not generated then we can enable it by default as it has not cost.
```suggestion
```

File: resources/scripts/frontend/spritesvg.js [hunk 16]
```diff
@@ -0,0 +1,3 @@
+const icons = require.context('@images/sprite-svg', true, /\.svg$/);
```
Comment (atanas-dev) [inline resources/scripts/frontend/spritesvg.js:1]: This seems like it would require the user to restart their build process when they add a new svg file to the sprite directory for it to be detected.
A watcher plugin (e.g. something based on `chokidar`) should be added to re-trigger a build on file add/delete/change/etc. events. It would even be better if we can keep this logic out of the frontend bundle and keep it on the webpack side entirely.

File: views/partials/loop.php [hunk 17]
```diff
@@ -20,6 +20,9 @@
 
 					<h2 class="article__title">
 						<a href="<?php the_permalink(); ?>" rel="bookmark" title="<?php echo esc_attr( app_get_permalink_title() ); ?>">
+							<svg class="icon icon-angle-right" width="1em" height="1em" fill="currentColor">
```
Comment (atanas-dev) [inline views/partials/loop.php]: This test change should be removed from the PR.

File: yarn.lock [hunk 18]
```diff
@@ -0,0 +1,11940 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (atanas-dev) [inline yarn.lock]: This file is intentionally not included with the theme and should not be included in this PR.

File: resources/build/spritesvg.js [hunk 19]
```diff
@@ -0,0 +1,12 @@
+/**
+ * The external dependencies.
+ */
+const SvgSpriteLoaderPlugin = require('svg-sprite-loader/plugin');
+
+module.exports = new SvgSpriteLoaderPlugin({
+  plainSprite: true,
+  spriteAttrs: {
+    'aria-hidden': 'true',
+    style: 'position: absolute; width: 0; height: 0; overflow: hidden;',
```
Comment (atanas-dev) [inline resources/build/spritesvg.js]: Is there any reason why we can't use `display: none; visibility: hidden;` here instead of absolute positioning?

File: resources/build/webpack.development.js [hunk 20]
```diff
@@ -144,6 +147,34 @@ module.exports = {
...
+       * Handle svg sprites.
+       */
+      {
+        test: utils.tests.svgs,
+        use: [
+          {
+            loader: 'svg-sprite-loader',
+            options: {
+              extract: true,
+              spriteFilename: 'images/sprite-svg.svg',
```
Comment (atanas-dev) [inline resources/build/webpack.development.js]:
```suggestion
              spriteFilename: 'images/sprite.svg',
```

**Review by atanas-dev — CHANGES_REQUESTED** [review]
A few adjustments are needed but overall a great start for this PR!

**Comment by paulomfr** [comment]
Hi @atanas-dev thank you by your reply, I'll look at this in the weekend.

File: resources/build/webpack.development.js (hunk 11 above)
Comment (paulomfr) [reply resources/build/webpack.development.js]: Hi @atanas-dev I did a merge with latest changes in my branch, and now I'm getting this error message:

Screen Shot 2020-04-23 at 07 56 15

Comment (atanas-dev) [reply resources/build/webpack.development.js]: You'll need to run `composer update` to get the latest master of all packages - I'm pushing updates often these days.

Comment (paulomfr) [reply resources/build/webpack.development.js]: Screen Shot 2020-04-27 at 07 12 29

After update, was returned this error.

Comment (atanas-dev) [reply resources/build/webpack.development.js]: You'll need to run `composer update` again.

File: resources/scripts/frontend/spritesvg.js (hunk 16 above)
Comment (paulomfr) [reply resources/scripts/frontend/spritesvg.js:1]: Hi @atanas-dev I've tested here and when add/delete/chante/etc happens the webpack will update the `sprite.svg`.

File: resources/build/webpack.development.js (hunk 13 above)
Comment (paulomfr) [reply resources/build/webpack.development.js]: In really, was a preset that I had used in a personal project, this was removed in the PR.

File: resources/build/spritesvg.js (hunk 19 above)
Comment (paulomfr) [reply resources/build/spritesvg.js]: In really, no reason, is that because in icomoon give me this set of style, but was removed until this is not really necessary.

**Comment by paulomfr** [comment]
Hi @atanas-dev I'm sorry for a while that Ii not made the changes, was a litte hard here. I've updated all changes in the PR with the new features of WPEmerge.

**Comment by paulomfr** [comment]
@atanas-dev can you take a look at my PR?

File: resources/build/lib/utils.js [hunk 21]
```diff
@@ -53,8 +53,9 @@ module.exports.distFontsPath = destPath =>
 module.exports.tests = {
   scripts: /\.(js|jsx)$/,
   styles: /\.(css|scss|sass)$/,
-  images: /(resources|dist|node_modules)[\\/].*\.(ico|jpg|jpeg|png|svg|gif)$/,
-  fonts: /(resources|dist|node_modules)[\\/].*\.(eot|svg|ttf|woff|woff2)$/,
+  svgs: /(resources|dist|node_modules)[\\/]images[\\/]sprite-svg[\\/].*\.svg$/,
+  images: /(resources|dist|node_modules)[\\/](?!images\/sprite-svg).*\.(ico|jpg|jpeg|png|svg|gif)$/,
+  fonts: /(resources|dist|node_modules)[\\/](?!images\/sprite-svg).*\.(eot|svg|ttf|woff|woff2)$/,
```
Comment (atanas-dev) [inline resources/build/lib/utils.js]: Minor adjustment to support Windows directory separators.
```suggestion
  images: /(resources|dist|node_modules)[\\/](?!images[\\/]sprite-svg).*\.(ico|jpg|jpeg|png|svg|gif)$/,
  fonts: /(resources|dist|node_modules)[\\/](?!images[\\/]sprite-svg).*\.(eot|svg|ttf|woff|woff2)$/,
```

File: composer.json (hunk 6 above)
Comment (atanas-dev) [reply composer.json]: > All changes to this file are not relevant to the PR and should be removed from it.

The PR still includes these changes.

File: resources/build/lib/utils.js [hunk 22]
```diff
@@ -53,8 +53,9 @@ module.exports.distFontsPath = destPath =>
 module.exports.tests = {
   scripts: /\.(js|jsx)$/,
   styles: /\.(css|scss|sass)$/,
-  images: /(resources|dist|node_modules)[\\/].*\.(ico|jpg|jpeg|png|svg|gif)$/,
-  fonts: /(resources|dist|node_modules)[\\/].*\.(eot|svg|ttf|woff|woff2)$/,
+  svgs: /(resources|dist|node_modules)[\\/]images[\\/]sprite-svg[\\/].*\.svg$/,
```
Comment (atanas-dev) [inline resources/build/lib/utils.js]:
```suggestion
  spriteSvgs: /(resources|dist|node_modules)[\\/]images[\\/]sprite-svg[\\/].*\.svg$/,
```

File: resources/build/webpack.development.js [hunk 23]
```diff
@@ -153,6 +155,22 @@ module.exports = {
         ],
       },
 
+      /**
+       * Handle SVG sprites.
+       */
+      {
+        test: utils.tests.svgs,
```
Comment (atanas-dev) [inline resources/build/webpack.development.js]:
```suggestion
        test: utils.tests.spriteSvgs,
```

File: resources/build/webpack.production.js [hunk 24]
```diff
@@ -193,6 +195,22 @@ module.exports = {
         ],
       },
 
+      /**
+       * Handle SVG sprites.
+       */
+      {
+        test: utils.tests.svgs,
```
Comment (atanas-dev) [inline resources/build/webpack.production.js]:
```suggestion
        test: utils.tests.spriteSvgs,
```

File: resources/build/webpack.production.js [hunk 25]
```diff
@@ -202,7 +220,7 @@ module.exports = {
           {
             loader: 'file-loader',
             options: {
-              name: file => `fonts/[name].${utils.filehash(file).substr(0, 10)}.[ext]`,
+              name: (file) => `fonts/[name].${utils.filehash(file).substr(0, 10)}.[ext]`,
```
Comment (atanas-dev) [inline resources/build/webpack.production.js]: This change does not seem necessary or related to SVG sprites - we should skip it from the PR.

File: resources/build/webpack.development.js [hunk 26]
```diff
@@ -153,6 +155,22 @@ module.exports = {
...
+        use: [
+          {
+            loader: 'svg-sprite-loader',
+            options: {
+              extract: true,
+              spriteFilename: 'images/sprite.svg',
```
Comment (atanas-dev) [inline resources/build/webpack.development.js]: Extract mode should not be used in webpack development/hot mode in order to have the svg sprite load correctly.
```suggestion
              extract: false,
```

File: resources/scripts/frontend/spritesvg.js [hunk 27]
```diff
@@ -0,0 +1,3 @@
...
+
+icons.keys().forEach(filename => icons(filename));
```
Comment (atanas-dev) [inline resources/scripts/frontend/spritesvg.js:4]:
```suggestion
// Automatically load all SVG files in the sprite-svg directory.
icons.keys().forEach(filename => icons(filename));
```

**Review by atanas-dev — CHANGES_REQUESTED** [review]
@paulomfr Good job - just a few more adjustments and this will be ready to merge :+1:

Once you're ready with the changes, please squash your commits into a single one with a short descriptive commit message.

**Comment by paulomfr** [comment]
@atanas-dev all changes done. I tried to use the `master` version to initialize a project but I'm having errors from compose. Is it unstable yet?

**Comment by atanas-dev** [comment]
Initializing `dev-master` via `composer create-project` won't work due to `dev-master` being required for all `wpemerge/*` packages as well. As long as you use the updated `CONTRIBUTING.md` steps you should be fine (via `git clone` rather than `composer create-project`).

I'll do my best to review the PR in the next few days.

**Comment by atanas-dev** [comment]
Thank you for the contribution, @paulomfr !
There are a couple details left but I'll take care of them in master ;)
